### PR TITLE
Basic support for message edits

### LINKED ISSUE #60: Bridge Message Edits
screenshot_20170930-083045

This is how the gitter bridge shows message edits without the ability to edit messages on the matrix side and I thought it was *gorgeous*. The repo for that bridge is [here](https://github.com/matrix-org/matrix-appservice-gitter).

I do not believe you can edit webhook messages, but once/if matrix gets message editing capabilities, we can send a similar message on the discord side, however perhaps with italics rather than colors, as discord is monocolor.

### PR DESCRIPTION
**Pull request by anoadragon453**
Needs some tests and perhaps some nicer formatting. But is functional at the moment.

image

image
--

For the sake of auditing, this fixes #60 - *Half-Shot*

### CODE REVIEW

File: src/bot.ts [hunk 1]
```diff
@@ -586,6 +587,36 @@ export class DiscordBot {
     });
   }
 
+  private async OnUpdate(oldMsg: Discord.Message, newMsg: Discord.Message) {
+    // Check if an edit was actually made
+    if (oldMsg.toString() === newMsg.toString()) {
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply src/bot.ts]: Opting for `.content` instead.

**Comment by anoadragon453** [comment]
This new SendMessage method is going to need a discord channel, author, messageID and guildID. All of this is contained in a `Discord.Message` object, so do we want to make a method like:

```
private SendMatrixMessageWithDiscordContext(discordMsg: Discord.Message, processorResult: MessageProcessorMatrixResult) {
```

or does that seem too messy? The method would look like:

```
  private async SendMessageWithDiscord(discordMsg: Discord.Message, matrixMsg: MessageProcessorMatrixResult) {
    const rooms = await this.GetRoomIdsFromChannel(discordMsg.channel)
    const intent = this.GetIntentFromDiscordMember(discordMsg.author);

    rooms.forEach((room) => {
      intent.sendMessage(room, {
        body: matrixMsg.body,
        msgtype: "m.text",
        formatted_body: matrixMsg.formattedBody,
        format: "org.matrix.custom.html",
      }).then((res) => {
        const evt = new DbEvent();
        evt.MatrixId = res.event_id + ";" + room;
        evt.DiscordId = discordMsg.id;
        evt.ChannelId = discordMsg.channel.id;
        evt.GuildId = discordMsg.guild.id;
        this.store.Insert(evt);
      });
    });
  }
```

**Comment by anoadragon453** [comment]
However, markdown is not formatted, because if I do so, `<p>` tags end up in the front and back, messing up the `*edited*` message:

image

Any way to not have leading and ending `<p>` tags to not be inserted?

**Comment by Half-Shot** [comment]
Sorry for the delay, I'm back and looking at this.

**Comment by Half-Shot** [comment]
> Any way to not have leading and ending <p> tags inserted?

Your best bet really is to not do the markdown-> HTML conversion until you have concatenated both the before and after messages.

**Comment by anoadragon453** [comment]
Sounds good. I'll try that and add a test for it.

File: src/bot.ts [hunk 2]
```diff
@@ -434,6 +435,32 @@ export class DiscordBot {
...
+        evt.MatrixId = res.event_id + ";" + room;
+        evt.DiscordId = msgID;
+        evt.ChannelId = chan.id;
+        evt.GuildId = guild.id;
+        this.store.Insert(evt);
+      });
+    });
+
+    // Sending was a success
+    return Promise.resolve(true);
```
Comment (Half-Shot) [inline src/bot.ts]: You can probably get away with ``return true;`` since it's ``async``.

File: src/bot.ts [hunk 3]
```diff
@@ -586,6 +613,19 @@ export class DiscordBot {
...
+    // Check if an edit was actually made
+    if (oldMsg.content === newMsg.content) {
+      return;
+    }
+
+    // Create a new edit message using the old and new message contents
+    const editedMsg = await this.msgProcessor.FormatEdit(oldMsg, newMsg);
+
+    // Send the message to all bridged matrix rooms
+    this.SendMatrixMessage(editedMsg, newMsg.channel, newMsg.guild, newMsg.author, newMsg.id);
```
Comment (Half-Shot) [inline src/bot.ts]: This should return

File: src/messageprocessor.ts [hunk 4]
```diff
@@ -71,6 +71,12 @@ export class MessageProcessor {
         return result;
     }
 
+    public async FormatEdit(oldMsg: Discord.Message, newMsg: Discord.Message): Promise<MessageProcessorMatrixResult> {
+        // TODO: Produce a nice, colored diff between the old and new message content
+        oldMsg.content = "*edit:* ~~" + oldMsg.content + "~~ -> " + newMsg.content;
+        return await this.FormatDiscordMessage(oldMsg);
```
Comment (Half-Shot) [inline src/messageprocessor.ts]: This doesn't need to await

**Comment by Half-Shot** [comment]
Okay, this all looks fine! :tada:  :taco:  :tanabata_tree:

**Pull request merged**